### DISCUSSION TITLE
How to analyze SPARQL execution plans using Jena 4.7.0

### DISCUSSION THREAD

**Discussion by chenkuncloud** [apache/jena#1808] [Q&A] [answered]
Hello,

How to analyze SPARQL execution plans using Jena 4.7.0, could you give me a complete code sample?

By the way, can Jena analyze the execution plan for SPARQL federated queries? For example:
```text
SELECT * WHERE {
    SERVICE <https://dbpedia.org/sparql> {
        ?s ?p ?o .
        FILTER ( ?o = 'water'@en )
    }
}
```

**Comment by kinow** [comment, marked as answer]
Hi @chenkun19960427 have you seen this page in the docs already? https://jena.apache.org/documentation/query/explain.html

It contains some information about the ARQ query parse explain command, `arq.qparse --explain 'SELECT * { ?s ?p ?o }'`. See if that helps :+1:

**Reply by chenkuncloud** [reply]
Hi, @kinow  I got the analysis results through the tutorial, thank you!

Java code:
```text
Model model = ModelFactory.createDefaultModel();
try (QueryExecution qExec = QueryExecution.create().model(model).query("SELECT * WHERE {\n" +
        "    SERVICE <https://dbpedia.org/sparql> {\n" +
        "        ?s ?p ?o .\n" +
        "        FILTER ( ?o = 'water'@en )\n" +
        "    }\n" +
        "}").set(ARQ.symLogExec, Explain.InfoLevel.ALL).build()) {
    ResultSet rs = qExec.execSelect();
}
```

Output:
```text
[main] INFO org.apache.jena.arq.exec - QUERY
  SELECT  *
  WHERE
    { SERVICE <https://dbpedia.org/sparql>
        { ?s  ?p  ?o
          FILTER ( ?o = "water"@en )
        }
    }
[main] INFO org.apache.jena.arq.exec - ALGEBRA
  (service <https://dbpedia.org/sparql>
    (filter (= ?o "water"@en)
      (bgp (triple ?s ?p ?o))))
```

**Comment by rvesse** [comment]
> By the way, can Jena analyze the execution plan for SPARQL federated queries?

No

If you analyse the query as a whole you'll see Jena has a `service` operator for evaluating the `SERVICE` portion of the query.  Everything within that operator is passed to the remote endpoint mostly as-is (possibly with some value substitution to restrict the remote query where appropriate) and how the remote endpoint chooses to evaluate its portion of the query is entirely beyond Jena's control

**Reply by chenkuncloud** [reply]
Hi, @rvesse  does that mean Jena Explain only works on local TDB databases, not on any remote endpoints (Virtuoso, Fuseki)?

**Comment by chenkuncloud** [comment]
@kinow Sorry, by the way, Is there any information on how to interpret the EXPLAIN results?

**Reply by kinow** [reply]
Hmm, there's some info about the SSE (SPARQL Syntax Expressions) output here: 

- https://jena.apache.org/documentation/notes/sse.html
- https://en.wikipedia.org/wiki/SPARQL_Syntax_Expressions

And you can find some presentations/videos/docs if you search for "bgp" "algebra" "sparql" "jena", for instance:

- http://events17.linuxfoundation.org/sites/events/files/slides/SPARQL%20Optimisation%20101%20Tutorial.pdf

Some commands may have been renamed or its syntaxes may have changed, but you should be able to follow it after reading Jena's docs and trying some basic queries.

**Reply by chenkuncloud** [reply]
Thank you. These are very useful!
